### DISCUSSION TITLE
Enable logging after trigger

### DISCUSSION THREAD

**Discussion by danbone** [sparcians/map#447] [Q&A] [answered]
Is there a way I could enable logging only once a certain condition happens i.e. enable logging when the retired instruction counter passes 100k.

**Comment by ghost** [comment, marked as answer]
Yes!  In the [Olympia](https://github.com/riscv-software-src/riscv-perf-model) Performance Simulator (and in the Sparta CoreModel), you can use the command line option `--debug-on-icount 100k` to start whatever collection mechanisms you've defined on the command line.  For example:

```
olympia -l top info my_info_log.out --debug-on-icount 100k traces/dhry_riscv.zstf 
```
This will start collecting a log on the `top` node after 100k instructions have been retired.

How this works...

Since the command line option is a _Sparta_ command line option, Sparta needs to know which counter in the implemented simulator that represents the `icount`.  This is communicated to Sparta via the derived `sparta::app::Simulation` instance overriding the method [`findSemanticCounter_`](https://github.com/sparcians/map/blob/f7a12a03947d5b4a174cc21a446fb76a8b9b1b36/sparta/sparta/app/Simulation.hpp#L650-L660)

For example:
```
    class MySim : public sparta::app::Simulation
    {
        // lots of stuff
    private:
        const sparta::CounterBase* findSemanticCounter_(CounterSemantic sem) const override {
             switch(sem){
             case CSEM_INSTRUCTIONS:
                return getRoot()->getChildAs<const sparta::CounterBase>("cpu.core0.rob.stats.total_number_retired");
                break;
             default:
                return nullptr;
        }
    };
```

**Comment by ghost** [comment]
Keep in mind that the command line options `--debug-on <cycle>` and `--debug-on-icount <inst count>` work on logging, pipeout collection, and pevent collection.

**Comment by danbone** [comment]
Thank you Knute, exactly what I was looking for. Can't believe I missed it in the help message.

**Reply by klingaard** [reply]
No prob.  This needed to be documented better.
